### DISCUSSION TITLE
Building simple UI within wezterm

### DISCUSSION THREAD

**Discussion by pombadev** [wezterm/wezterm#4867] [Ideas]
We have following components in wezterm:

![Screenshot from 2024-01-26 23-06-59](https://github.com/wez/wezterm/assets/11157420/738ac825-f51c-47c9-b13a-94cb9e16808d)

![Screenshot from 2024-01-26 23-06-41](https://github.com/wez/wezterm/assets/11157420/a47d4cd0-0e10-45e4-be08-ec189e6e36e7)

I was wondering if this is something that interest you to expose lua api to build a [dmenu](https://tools.suckless.org/dmenu/) like ui within wezterm.

My use case is:

- Select a text in the terminal and invoke a menu to perform various actions on that selection, eg:
   - Google that text with default browser
   - Open a url from one of the available browser
   - Copy and paste it back in terminal
   - etc
  
- VSCode like theme picker

Something like this
![Screenshot from 2024-01-26 23-31-13](https://github.com/wez/wezterm/assets/11157420/a43e4456-b234-454b-b687-14e40cddc10b)

**Comment by wez** [comment]
Take a look at https://wezfurlong.org/wezterm/config/lua/keyassignment/InputSelector.html

**Reply by pombadev** [reply]
Thank you for your hard work!, currently i have cobbled up these with but it covers up the entire terminal window.

- InputSelector
![image](https://github.com/wez/wezterm/assets/11157420/7ce96569-d4ac-4d61-b189-2b8058cd8279)

- PromptInputLine
![image](https://github.com/wez/wezterm/assets/11157420/33fa9686-d69b-4115-9939-9177e831268f)

But IMHO the `Modal` component looks great.

Feel free to mark this discussion as answered.

**Reply by GV14982** [reply]
I agree that having access to the sort of "floating pane" options would be great. Any idea how difficult that might be? I could be down to work on it. I see #5576 is an open PR. I suppose we could also just use this when it's done?

**Comment by basus** [comment]
I came across this discussion while trying to figure out if I could build something like Emacs' [which-key](https://github.com/justbur/emacs-which-key) or Helix's help popups. I think that once [floating panes](https://github.com/wez/wezterm/pull/5576) are merged, we could rebuild the modals and overlays in terms of floating panes. That would help them all have a more uniform look-and-feel.
